tokio-util/src/net/unix/mod.rs

```diff
@@ -0,0 +1,108 @@
+//! Unix domain socket helpers.
+
+use std::future::Future;
+use std::io::Result;
+use std::pin::Pin;
+
+/// Listening address.
+#[derive(Debug)]
+pub enum ListenAddr {
+    /// Socket address.
+    SocketAddr(std::net::SocketAddr),
+    /// Unix socket.
+    UnixSocket(tokio::net::unix::SocketAddr),
+}
+
+impl From<std::net::SocketAddr> for ListenAddr {
+    fn from(addr: std::net::SocketAddr) -> Self {
+        Self::SocketAddr(addr)
+    }
+}
+
+impl From<tokio::net::unix::SocketAddr> for ListenAddr {
+    fn from(addr: tokio::net::unix::SocketAddr) -> Self {
+        Self::UnixSocket(addr)
+    }
+}
+
+/// A trait for a listener: `TcpListener` and `UnixListener`.
+pub trait Listener: Send + Unpin {
+    /// Accepts a new incoming connection from this listener.
+    fn accept<'a>(
+        &'a self,
+    ) -> Pin<
+        Box<
+            dyn Future<
+                    Output = Result<(Box<dyn AsyncReadWrite + Send + Unpin + 'static>, ListenAddr)>,
+                > + Send
+                + 'a,
+        >,
+    >;
+
+    /// Returns the local address that this listener is bound to.
+    fn local_addr(&self) -> Result<ListenAddr>;
+}
+
+impl Listener for tokio::net::TcpListener {
+    fn accept<'a>(
+        &'a self,
+    ) -> Pin<
+        Box<
+            dyn Future<
+                    Output = Result<(Box<dyn AsyncReadWrite + Send + Unpin + 'static>, ListenAddr)>,
+                > + Send
+                + 'a,
+        >,
+    > {
+        let accept = self.accept();
+        Box::pin(async {
+            let (stream, addr) = accept.await?;
+            Ok((Box::new(stream) as Box<_>, addr.into()))
+        })
+    }
+
+    fn local_addr(&self) -> Result<ListenAddr> {
+        self.local_addr().map(Into::into)
+    }
+}
+
+impl Listener for tokio::net::UnixListener {
+    fn accept<'a>(
+        &'a self,
+    ) -> Pin<
+        Box<
+            dyn Future<
+                    Output = Result<(Box<dyn AsyncReadWrite + Send + Unpin + 'static>, ListenAddr)>,
+                > + Send
+                + 'a,
+        >,
+    > {
+        let accept = self.accept();
+        Box::pin(async {
+            let (stream, addr) = accept.await?;
+            Ok((Box::new(stream) as Box<_>, addr.into()))
+        })
+    }
+
+    fn local_addr(&self) -> Result<ListenAddr> {
+        self.local_addr().map(Into::into)
+    }
+}
+
+/// A trait that combines `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+pub trait AsyncReadWrite: tokio::io::AsyncRead + tokio::io::AsyncWrite {
+    /// Sets the value of the `TCP_NODELAY` option on this socket if this is a TCP socket.
+    fn set_nodelay(&self, nodelay: bool) -> Result<()>;
+}
```

Comment:
The name of this trait doesn't make sense. It's not a trait for combining `AsyncRead` and `AsyncWrite` - its a trait for adding an `set_nodelay` method to a type.

Reply:
This should probably be called...`SetNodelay`, and _not_ require `AsyncRead` and `AsyncWrite`...

Reply:
It's used for boxing.
```
Box<dyn AsyncReadWrite + Send + Unpin + 'static>
```
You can't add 2 traits in a Box dyn like that. You have to create a third one that inherit the 2 others.

`set_nodelay` here is detail in the trait. It's there because I needed it from TcpListener but other functions could be there too

Reply:
> You can't add 2 traits in a Box dyn like that. You have to create a third one that inherit the 2 others.

Right...I think that the boxed trait object should probably be replaced with an associated type like
```rust
type Io = tokio::io::AsyncRead + tokio::io::AsyncWrite;
```
as Alice mentioned here https://github.com/tokio-rs/tokio/pull/4244#discussion_r752463205. That way, user code _could_ define a `SetNodelay` trait, implement it for `TcpStream` and `UnixStream`, and have trait bounds like
```rust
async fn do_something_with_a_listener<L>(listener: L)
where
    L::Io: SetNodelay,
{
    // ...
}
```
This way, if other users need to expose other functionality on sockets, they can add similar traits and still use the `Listener` trait, rather than special-casing `set_nodelay` just because _your_ application needs to disable Nagle's algorithm.

Reply:
Yes, this should be an associated type, and the `set_nodelay` method should honestly just get taken out. It's too specific to your particular use-case.